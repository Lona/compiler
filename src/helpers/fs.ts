import * as fs from 'fs'
import * as path from 'path'

export type FSWrapper = {
  readFile(filePath: string): Promise<string>
  writeFile(filePath: string, data: string): Promise<void>
  copyDir(dirPath: string, output?: string): Promise<void>
}

export const createFSWrapper = (
  workspacePath: string,
  _outputPath?: unknown
): FSWrapper => {
  const outputPath =
    typeof _outputPath === 'string'
      ? _outputPath
      : path.join(process.cwd(), 'lona-generated')
  const fsWrapper = {
    readFile(filePath: string) {
      return fs.promises.readFile(
        path.resolve(workspacePath, filePath),
        'utf-8'
      )
    },
    writeFile(filePath: string, data: string) {
      const resolvedPath = path.resolve(outputPath, filePath)
      fs.mkdirSync(path.dirname(resolvedPath), { recursive: true })
      return fs.promises.writeFile(resolvedPath, data, 'utf-8')
    },
    async copyDir(dirPath: string, output: string = '.') {
      const resolvedPath = path.resolve(workspacePath, dirPath)
      const files = await fs.promises.readdir(resolvedPath)

      await Promise.all(
        files.map(async x => {
          if (
            (await fs.promises.stat(path.join(resolvedPath, x))).isDirectory()
          ) {
            return
          }

          return fsWrapper.writeFile(
            path.join(output, x),
            await fsWrapper.readFile(path.join(dirPath, x))
          )
        })
      )
    },
  }

  return fsWrapper
}
