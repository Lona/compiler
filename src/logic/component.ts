import { LogicAST as AST } from '@lona/serialization'
import { FunctionDeclaration } from './nodes/FunctionDeclaration'
import { findNode } from './traversal'
import { FunctionCallExpression } from './nodes/FunctionCallExpression'
import { IdentifierExpression } from './nodes/IdentifierExpression'
import { IExpression } from './nodes/interfaces'
import { ComponentVisitor } from './components/ComponentVisitor'
import { LonaView, LonaViewConstructor } from './components/LonaView'
import { LonaText } from './components/LonaText'
import { LonaStack } from './components/LonaStack'

export type ComponentContext = {
  intrinsicNameCount: { [key: string]: number }
  viewAttributeAssignment: {
    [key: string]: { [key: string]: IExpression }
  }
}

export function createViewHierarchy(
  visitor: ComponentVisitor,
  node: IExpression
): LonaView {
  if (node instanceof FunctionCallExpression) {
    const { callee, argumentExpressionNodes } = node

    if (!(callee instanceof IdentifierExpression)) {
      throw new Error('Element function must be an identifier (for now)')
    }

    switch (callee.name) {
      case 'View': {
        return new LonaView(node as LonaViewConstructor, visitor)
      }
      case 'VerticalStack':
      case 'HorizontalStack': {
        return new LonaStack(node as LonaViewConstructor, visitor)
      }
      case 'Text': {
        return new LonaText(node as LonaViewConstructor, visitor)
      }
      default:
        throw new Error(`Unknown element name: ${callee.name}`)
    }
  }

  throw new Error('Unhandled element')
}

export function findComponentFunction(
  node: AST.SyntaxNode
): FunctionDeclaration | undefined {
  const syntaxNode = findNode(node, child => child.type === 'function') as
    | AST.FunctionDeclaration
    | undefined

  if (syntaxNode) {
    const node = new FunctionDeclaration(syntaxNode)
    return node
  }
}
