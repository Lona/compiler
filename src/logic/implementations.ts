import { FuncImplementation } from './runtime/memory'
import { Decode, Encode } from './runtime/value'
import Color from 'color'

const implementations: {
  [key: string]: FuncImplementation
} = {
  'Color.saturate': ({ color, factor }) => {
    const colorString = Decode.color(color)
    const factorNumber = Decode.number(factor) ?? 1

    const result = Color(colorString)
      .saturate(factorNumber)
      .hex()

    return Encode.color(result).memory
  },
}

export default implementations
