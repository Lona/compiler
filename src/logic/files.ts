import fs from 'fs'
import { IFS, match } from 'buffs'
import path from 'path'

export function componentFilePaths(fs: IFS, workspacePath: string): string[] {
  return match(fs, workspacePath, { includePatterns: ['**/*.cmp'] }).map(file =>
    path.join(workspacePath, file)
  )
}

export function logicFilePaths(fs: IFS, workspacePath: string): string[] {
  return match(fs, workspacePath, {
    includePatterns: ['**/*.logic'],
  }).map(file => path.join(workspacePath, file))
}

export function documentFilePaths(fs: IFS, workspacePath: string): string[] {
  return match(fs, workspacePath, { includePatterns: ['**/*.md'] }).map(file =>
    path.join(workspacePath, file)
  )
}

export function libraryFilePaths(): string[] {
  const libraryPath = path.join(__dirname, 'library')

  return match(fs, libraryPath, { includePatterns: ['**/*.logic'] }).map(file =>
    path.join(libraryPath, file)
  )
}

export function decode<T>(
  fs: IFS,
  filePath: string,
  decoder: (data: string) => T
): T {
  const data = fs.readFileSync(filePath, 'utf8')
  const ast = decoder(data)
  return ast
}
