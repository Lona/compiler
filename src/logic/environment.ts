import * as fs from 'fs'
import * as path from 'path'
import { LogicAST as AST, decodeLogic } from '@lona/serialization'
import { unify } from './typeUnifier'
import * as LogicEvaluate from './evaluation'
import { createNamespace } from './namespace'
import { createUnificationContext } from './typeChecker'
import { createScopeContext } from './scope'
import { Reporter } from '../utils/reporter'
import { makeProgram, joinPrograms } from '../helpers/logicAst'
import { nonNullable } from '../utils'

export const STANDARD_LIBRARY = 'standard library'

export const run = (reporter: Reporter, programs: AST.SyntaxNode[]) => {
  const standardLibsPath = path.join(__dirname, 'library')
  const standardLibs = fs.readdirSync(standardLibsPath)

  const libraryFiles: AST.Program[] = standardLibs.map(
    x =>
      makeProgram(
        decodeLogic(fs.readFileSync(path.join(standardLibsPath, x), 'utf8'))
      ) as AST.Program
  )

  const standardLibsProgram = joinPrograms(libraryFiles)
  const logicPrograms = programs.map(makeProgram).filter(nonNullable)
  const programNode = joinPrograms([standardLibsProgram, ...logicPrograms])

  const namespace = createNamespace(programNode)
  const scope = createScopeContext(programNode, namespace)
  const typeChecker = createUnificationContext(programNode, scope, reporter)
  const substitution = unify(typeChecker.constraints, reporter)

  const evaluationContext = LogicEvaluate.evaluate(
    programNode,
    scope,
    typeChecker,
    substitution,
    reporter
  )

  return evaluationContext
}
