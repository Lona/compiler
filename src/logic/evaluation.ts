import { LogicAST as AST } from '@lona/serialization'
import { EvaluationVisitor } from './evaluationVisitor'
import { createEvaluationVisitor } from './nodes/createNode'
import { Value } from './runtime/value'
import { Scope } from './scope'
import { TypeChecker } from './typeChecker'
import { Substitution } from './typeUnifier'
import { Reporter } from '../utils/reporter'
import { declarationPathTo } from './ast'
import { assertNever } from '../utils/typeHelpers'

const STANDARD_LIBRARY = 'standard library'

export function evaluateIsTrue(
  context: EvaluationContext,
  expression: AST.Expression
) {
  const condition = context.evaluate(expression.data.id)
  return (
    (condition &&
      condition.type.type === 'constructor' &&
      condition.type.name === 'Boolean' &&
      condition.memory.type === 'bool' &&
      condition.memory.value) ||
    false
  )
}

export type Thunk = {
  label: string
  dependencies: string[]
  f: (args: Value[]) => Value
}

/**
 * The evaluation context of the Lona Workspace.
 */
export class EvaluationContext {
  values: { [uuid: string]: Value } = {}
  thunks: { [uuid: string]: Thunk } = {}
  reporter?: Reporter

  constructor(reporter?: Reporter) {
    this.reporter = reporter
  }

  add(uuid: string, thunk: Thunk) {
    this.thunks[uuid] = thunk
  }

  addValue(uuid: string, value: Value) {
    this.values[uuid] = value
  }

  /**
   * Evaluate the id to a value, resolving any dependency along the way
   */
  evaluate(
    uuid: string,
    reporter: Reporter | undefined = this.reporter
  ): Value | undefined {
    const value = this.values[uuid]

    if (value) return value

    const thunk = this.thunks[uuid]

    if (!thunk) {
      reporter?.error(`no thunk for ${uuid}`)
      return undefined
    }

    const resolvedDependencies = thunk.dependencies.map(x =>
      this.evaluate(x, reporter)
    )

    if (resolvedDependencies.some(x => !x)) {
      reporter?.error(
        `Failed to evaluate thunk ${uuid} (${thunk.label}) - missing dep ${
          thunk.dependencies[resolvedDependencies.findIndex(x => !x)]
        }`
      )
      return undefined
    }

    const result = thunk.f(resolvedDependencies as Value[])
    this.values[uuid] = result

    return result
  }

  copy() {
    const newContext = new EvaluationContext(this.reporter)
    newContext.thunks = { ...this.thunks }
    newContext.values = { ...this.values }
    return newContext
  }
}

const evaluateNode = (
  node: AST.SyntaxNode,
  visitor: EvaluationVisitor
): EvaluationContext => {
  // TODO: Handle stopping
  const context = AST.subNodes(node).reduce<EvaluationContext>(
    (prev, subNode) => {
      return evaluateNode(subNode, visitor)
    },
    visitor.evaluation
  )

  if (!context) return context

  switch (node.type) {
    case 'identifierExpression':
    case 'none':
    case 'boolean':
    case 'number':
    case 'string':
    case 'color':
    case 'array':
    case 'literalExpression':
    case 'memberExpression':
    case 'record':
    case 'variable':
    case 'enumeration':
    case 'function':
    case 'functionCallExpression':
      const visitorNode = createEvaluationVisitor(node)

      if (visitorNode) {
        visitorNode.evaluationEnter(visitor)
      }

      break
    case 'assignmentExpression': {
      visitor.add(node.data.left.data.id, {
        label:
          'Assignment for ' +
          declarationPathTo(visitor.rootNode, node.data.left.data.id).join('.'),
        dependencies: [node.data.right.data.id],
        f: values => values[0],
      })
      break
    }
    case 'functionType':
    case 'typeIdentifier':
    case 'program':
    case 'parameter':
    case 'value':
    case 'topLevelParameters':
    case 'topLevelDeclarations':
    case 'enumerationCase': // handled in 'enumeration'
    case 'argument': // handled in 'functionCallExpression'
    case 'namespace':
    case 'importDeclaration':
    case 'placeholder':
    case 'return': // handled in 'function'
    case 'loop': // handled in 'function'
    case 'branch': // handled in 'function'
    case 'expression':
    case 'declaration': {
      break
    }
    default: {
      assertNever(node)
    }
  }

  return context
}

export const evaluate = (
  rootNode: AST.SyntaxNode,
  scope: Scope,
  typeChecker: TypeChecker,
  substitution: Substitution,
  reporter: Reporter
): EvaluationContext => {
  const visitor = new EvaluationVisitor(
    rootNode,
    scope,
    typeChecker,
    substitution,
    reporter
  )

  return evaluateNode(rootNode, visitor)
}
