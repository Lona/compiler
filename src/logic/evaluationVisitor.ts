import { LogicAST as AST } from '@lona/serialization'
import { Value } from './runtime/value'
import { Scope } from './scope'
import { TypeChecker } from './typeChecker'
import { Substitution, substitute } from './typeUnifier'
import { EvaluationContext, Thunk } from './evaluation'
import { UUID, Namespace } from './namespace'
import { StaticType } from './staticType'
import { Reporter } from '../utils/reporter'

export class EvaluationVisitor {
  evaluation: EvaluationContext
  rootNode: AST.SyntaxNode
  scope: Scope
  reporter: Reporter
  typeChecker: TypeChecker
  substitution: Substitution
  namespace: Namespace

  constructor(
    rootNode: AST.SyntaxNode,
    namespace: Namespace,
    scope: Scope,
    typeChecker: TypeChecker,
    substitution: Substitution,
    reporter: Reporter
  ) {
    this.evaluation = new EvaluationContext(reporter)
    this.rootNode = rootNode
    this.namespace = namespace
    this.scope = scope
    this.reporter = reporter
    this.typeChecker = typeChecker
    this.substitution = substitution
  }

  add(uuid: UUID, thunk: Thunk) {
    this.evaluation.add(uuid, thunk)
  }

  addValue(uuid: UUID, value: Value) {
    this.evaluation.addValue(uuid, value)
  }

  resolveType = (uuid: UUID): StaticType | undefined => {
    const { typeChecker, substitution, reporter } = this

    let type = typeChecker.nodes[uuid]

    if (!type) {
      reporter.error(`Unknown type ${uuid}`)
      return
    }

    return substitute(substitution, type)
  }
}
