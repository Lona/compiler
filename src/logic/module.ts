import fs from 'fs'
import path from 'path'
import {
  decodeDocument,
  decodeLogic,
  extractProgramFromAST,
  LogicAST as AST,
  MDXAST,
} from '@lona/serialization'
import { IFS, match } from 'buffs'
import { evaluate, EvaluationContext } from './evaluation'
import { createNamespace, mergeNamespaces, Namespace } from './namespace'
import { createScopeContext, mergeScopes, Scope } from './scope'
import { createUnificationContext, TypeChecker } from './typeChecker'
import { Substitution, unify } from './typeUnifier'
import { joinPrograms, makeProgram } from './ast'

export type LogicFile = {
  isLibrary: boolean
  sourcePath: string
  rootNode: AST.TopLevelDeclarations
  mdxContent: MDXAST.Content[]
}

export type ModuleContext = {
  componentFiles: LogicFile[]
  libraryFiles: LogicFile[]
  documentFiles: LogicFile[]
  logicFiles: LogicFile[]
  sourceFiles: LogicFile[]
  namespace: Namespace
  scope: Scope
  typeChecker: TypeChecker
  substitution: Substitution
  evaluationContext: EvaluationContext
  isValidProgram: boolean
}

export function createModule(
  workspaceFs: IFS,
  workspacePath: string
): ModuleContext {
  const libraryFiles: LogicFile[] = libraryFilePaths().map(sourcePath => ({
    isLibrary: true,
    sourcePath,
    // Always read library files from the real FS
    rootNode: decodeLogic(
      fs.readFileSync(sourcePath, 'utf8')
    ) as AST.TopLevelDeclarations,
    mdxContent: [],
  }))

  const componentFiles: LogicFile[] = componentFilePaths(
    workspaceFs,
    workspacePath
  ).map(sourcePath => ({
    isLibrary: false,
    sourcePath,
    rootNode: decodeLogic(
      workspaceFs.readFileSync(sourcePath, 'utf8')
    ) as AST.TopLevelDeclarations,
    mdxContent: [],
  }))

  const logicFiles: LogicFile[] = logicFilePaths(
    workspaceFs,
    workspacePath
  ).map(sourcePath => ({
    isLibrary: false,
    sourcePath,
    rootNode: decodeLogic(
      workspaceFs.readFileSync(sourcePath, 'utf8')
    ) as AST.TopLevelDeclarations,
    mdxContent: [],
  }))

  const documentFiles: LogicFile[] = documentFilePaths(
    workspaceFs,
    workspacePath
  ).map(sourcePath => {
    const decoded = decodeDocument(workspaceFs.readFileSync(sourcePath, 'utf8'))

    return {
      isLibrary: false,
      sourcePath,
      rootNode: extractProgramFromAST(decoded),
      mdxContent: decoded.children,
    }
  })

  const files: LogicFile[] = [
    ...libraryFiles,
    ...componentFiles,
    ...documentFiles,
    ...logicFiles,
  ]

  const namespace: Namespace = mergeNamespaces(
    files.map(logicFile => createNamespace(logicFile.rootNode))
  )

  const scope: Scope = mergeScopes(
    files.map(logicFile =>
      createScopeContext(logicFile.rootNode, namespace, undefined, console)
    )
  )

  const programNode = joinPrograms(
    files.map(logicFile => makeProgram(logicFile.rootNode))
  )

  const typeChecker = createUnificationContext(programNode, scope, console)

  const substitution = unify(typeChecker.constraints, console)

  const evaluationContext = evaluate(
    programNode,
    namespace,
    scope,
    typeChecker,
    substitution,
    console
  )

  return {
    componentFiles,
    libraryFiles,
    documentFiles,
    logicFiles,
    get sourceFiles() {
      return [...documentFiles, ...logicFiles]
    },
    namespace,
    scope,
    typeChecker,
    substitution,
    evaluationContext,
    isValidProgram:
      scope.undefinedIdentifierExpressions.size === 0 &&
      scope.undefinedMemberExpressions.size === 0 &&
      scope.undefinedTypeIdentifiers.size === 0,
  }
}

function componentFilePaths(fs: IFS, workspacePath: string): string[] {
  return match(fs, workspacePath, { includePatterns: ['**/*.cmp'] }).map(file =>
    path.join(workspacePath, file)
  )
}

function logicFilePaths(fs: IFS, workspacePath: string): string[] {
  return match(fs, workspacePath, {
    includePatterns: ['**/*.logic'],
  }).map(file => path.join(workspacePath, file))
}

function documentFilePaths(fs: IFS, workspacePath: string): string[] {
  return match(fs, workspacePath, { includePatterns: ['**/*.md'] }).map(file =>
    path.join(workspacePath, file)
  )
}

export function libraryFilePaths(): string[] {
  return match(fs, STANDARD_LIBRARY_PATH, {
    includePatterns: ['**/*.logic'],
  }).map(file => path.join(STANDARD_LIBRARY_PATH, file))
}

export const STANDARD_LIBRARY_PATH = path.join(__dirname, '../../static/logic')
