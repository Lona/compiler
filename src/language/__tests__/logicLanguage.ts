import fs from 'fs'
import path from 'path'
// import { buildLexerFromSource, buildParserFromSource } from '../build'
// import { Printer } from '../Printer'
import { inspect } from 'util'
import { createModule, libraryFilePaths } from '../../logic/module'

const grammarSource = fs.readFileSync(
  path.join(__dirname, './mocks/LogicLanguage.logic'),
  'utf8'
)

describe('Logic Language', () => {
  it('empty', () => {})

  //   const lexer = buildLexerFromSource(grammarSource)
  //   const parser = buildParserFromSource(grammarSource)
  //   const printer = new Printer(lexer.stateDefinitions, parser.definition)

  //   it('tokenizes numbers', () => {
  //     const tokens = lexer.tokenize('-12.34')
  //     expect(printer.print(printer.formatTokens(tokens))).toEqual('-12.34')
  //   })

  //   it('tokenizes records', () => {
  //     const tokens = lexer.tokenize(`struct Foo {
  //   let foo: Number = 42
  // }`)
  //     expect(tokens).toMatchSnapshot()
  //   })

  //   it('tokenizes enums', () => {
  //     const tokens = lexer.tokenize(`enum Foo {
  //   case foo(a: Number, b: String)
  // }`)
  //     expect(tokens).toMatchSnapshot()
  //   })

  //   it('tokenizes functions', () => {
  //     const tokens = lexer.tokenize(`func test(a: Number = 123) -> Boolean {
  //   return true
  // }`)
  //     expect(tokens).toMatchSnapshot()
  //   })

  //   it('tokenizes comments', () => {
  //     const tokens = lexer.tokenize(`/* test1 */ let a = 123 /*test2*/`)
  //     expect(tokens).toMatchSnapshot()
  //   })

  //   it('tokenizes the standard library successfully', () => {
  //     libraryFilePaths().forEach(filePath => {
  //       const source = fs.readFileSync(filePath, 'utf8')
  //       lexer.tokenize(source)
  //     })
  //   })

  //   it('parses identifiers', () => {
  //     const root = 'LGCIdentifier'
  //     const tokens = lexer.tokenize(`foo`)
  //     const result = parser.parse(tokens, root)

  //     expect(result.type).toEqual('success')

  //     if (result.type === 'success') {
  //       expect(result.value).toMatchSnapshot()
  //       expect(printer.print(printer.formatNode(result.value, root))).toEqual(
  //         'foo'
  //       )
  //     }
  //   })

  //   it('parses type annotation', () => {
  //     const root = 'LGCTypeAnnotation'
  //     const tokens = lexer.tokenize(`Foo<A>`)
  //     const result = parser.parse(tokens, root)

  //     expect(result.type).toEqual('success')

  //     if (result.type === 'success') {
  //       expect(result.value).toMatchSnapshot()
  //       expect(
  //         printer.print(printer.formatNode(result.value, root))
  //       ).toMatchSnapshot()
  //     }
  //   })

  //   it('parses variable declarations', () => {
  //     const root = 'LGCVariableDeclaration'
  //     const tokens = lexer.tokenize(`let foo: Number =`)
  //     const result = parser.parse(tokens, root)

  //     expect(result.type).toEqual('success')

  //     if (result.type === 'success') {
  //       expect(result.value).toMatchSnapshot()
  //       expect(
  //         printer.print(printer.formatNode(result.value, root))
  //       ).toMatchSnapshot()
  //     }
  //   })
})
