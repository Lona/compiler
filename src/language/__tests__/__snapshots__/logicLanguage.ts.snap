// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logic Language parses identifiers 1`] = `
Object {
  "string": "foo",
}
`;

exports[`Logic Language parses type annotation 1`] = `
Object {
  "genericArguments": Array [
    Object {
      "genericArguments": Array [],
      "identifier": Object {
        "string": "A",
      },
    },
  ],
  "identifier": Object {
    "string": "Foo",
  },
}
`;

exports[`Logic Language parses type annotation 2`] = `"Foo"`;

exports[`Logic Language parses variable declarations 1`] = `
Object {
  "name": Object {
    "string": "foo",
  },
}
`;

exports[`Logic Language parses variable declarations 2`] = `"letfoo:"`;

exports[`Logic Language tokenizes comments 1`] = `
Array [
  Object {
    "groups": Array [
      " test1 ",
    ],
    "position": Object {
      "end": 11,
      "start": 0,
    },
    "type": "comment",
    "value": "/* test1 */",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 15,
      "start": 12,
    },
    "type": "let",
    "value": "let",
  },
  Object {
    "groups": Array [
      "a",
    ],
    "position": Object {
      "end": 17,
      "start": 16,
    },
    "type": "identifier",
    "value": "a",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 19,
      "start": 18,
    },
    "type": "eq",
    "value": "=",
  },
  Object {
    "groups": Array [
      "123",
    ],
    "position": Object {
      "end": 23,
      "start": 20,
    },
    "type": "integer",
    "value": "123",
  },
  Object {
    "groups": Array [
      "test2",
    ],
    "position": Object {
      "end": 33,
      "start": 24,
    },
    "type": "comment",
    "value": "/*test2*/",
  },
]
`;

exports[`Logic Language tokenizes enums 1`] = `
Array [
  Object {
    "groups": Array [],
    "position": Object {
      "end": 4,
      "start": 0,
    },
    "type": "enum",
    "value": "enum",
  },
  Object {
    "groups": Array [
      "Foo",
    ],
    "position": Object {
      "end": 8,
      "start": 5,
    },
    "type": "identifier",
    "value": "Foo",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 10,
      "start": 9,
    },
    "type": "openBrace",
    "value": "{",
  },
  Object {
    "groups": Array [
      "case",
    ],
    "position": Object {
      "end": 17,
      "start": 13,
    },
    "type": "identifier",
    "value": "case",
  },
  Object {
    "groups": Array [
      "foo",
    ],
    "position": Object {
      "end": 21,
      "start": 18,
    },
    "type": "identifier",
    "value": "foo",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 22,
      "start": 21,
    },
    "type": "openParens",
    "value": "(",
  },
  Object {
    "groups": Array [
      "a",
    ],
    "position": Object {
      "end": 23,
      "start": 22,
    },
    "type": "identifier",
    "value": "a",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 24,
      "start": 23,
    },
    "type": "colon",
    "value": ":",
  },
  Object {
    "groups": Array [
      "Number",
    ],
    "position": Object {
      "end": 31,
      "start": 25,
    },
    "type": "identifier",
    "value": "Number",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 32,
      "start": 31,
    },
    "type": "comma",
    "value": ",",
  },
  Object {
    "groups": Array [
      "b",
    ],
    "position": Object {
      "end": 34,
      "start": 33,
    },
    "type": "identifier",
    "value": "b",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 35,
      "start": 34,
    },
    "type": "colon",
    "value": ":",
  },
  Object {
    "groups": Array [
      "String",
    ],
    "position": Object {
      "end": 42,
      "start": 36,
    },
    "type": "identifier",
    "value": "String",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 43,
      "start": 42,
    },
    "type": "closeParens",
    "value": ")",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 45,
      "start": 44,
    },
    "type": "closeBrace",
    "value": "}",
  },
]
`;

exports[`Logic Language tokenizes functions 1`] = `
Array [
  Object {
    "groups": Array [],
    "position": Object {
      "end": 4,
      "start": 0,
    },
    "type": "func",
    "value": "func",
  },
  Object {
    "groups": Array [
      "test",
    ],
    "position": Object {
      "end": 9,
      "start": 5,
    },
    "type": "identifier",
    "value": "test",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 10,
      "start": 9,
    },
    "type": "openParens",
    "value": "(",
  },
  Object {
    "groups": Array [
      "a",
    ],
    "position": Object {
      "end": 11,
      "start": 10,
    },
    "type": "identifier",
    "value": "a",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 12,
      "start": 11,
    },
    "type": "colon",
    "value": ":",
  },
  Object {
    "groups": Array [
      "Number",
    ],
    "position": Object {
      "end": 19,
      "start": 13,
    },
    "type": "identifier",
    "value": "Number",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 21,
      "start": 20,
    },
    "type": "eq",
    "value": "=",
  },
  Object {
    "groups": Array [
      "123",
    ],
    "position": Object {
      "end": 25,
      "start": 22,
    },
    "type": "integer",
    "value": "123",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 26,
      "start": 25,
    },
    "type": "closeParens",
    "value": ")",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 29,
      "start": 27,
    },
    "type": "arrow",
    "value": "->",
  },
  Object {
    "groups": Array [
      "Boolean",
    ],
    "position": Object {
      "end": 37,
      "start": 30,
    },
    "type": "identifier",
    "value": "Boolean",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 39,
      "start": 38,
    },
    "type": "openBrace",
    "value": "{",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 48,
      "start": 42,
    },
    "type": "return",
    "value": "return",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 53,
      "start": 49,
    },
    "type": "true",
    "value": "true",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 55,
      "start": 54,
    },
    "type": "closeBrace",
    "value": "}",
  },
]
`;

exports[`Logic Language tokenizes records 1`] = `
Array [
  Object {
    "groups": Array [],
    "position": Object {
      "end": 6,
      "start": 0,
    },
    "type": "struct",
    "value": "struct",
  },
  Object {
    "groups": Array [
      "Foo",
    ],
    "position": Object {
      "end": 10,
      "start": 7,
    },
    "type": "identifier",
    "value": "Foo",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 12,
      "start": 11,
    },
    "type": "openBrace",
    "value": "{",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 18,
      "start": 15,
    },
    "type": "let",
    "value": "let",
  },
  Object {
    "groups": Array [
      "foo",
    ],
    "position": Object {
      "end": 22,
      "start": 19,
    },
    "type": "identifier",
    "value": "foo",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 23,
      "start": 22,
    },
    "type": "colon",
    "value": ":",
  },
  Object {
    "groups": Array [
      "Number",
    ],
    "position": Object {
      "end": 30,
      "start": 24,
    },
    "type": "identifier",
    "value": "Number",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 32,
      "start": 31,
    },
    "type": "eq",
    "value": "=",
  },
  Object {
    "groups": Array [
      "42",
    ],
    "position": Object {
      "end": 35,
      "start": 33,
    },
    "type": "integer",
    "value": "42",
  },
  Object {
    "groups": Array [],
    "position": Object {
      "end": 37,
      "start": 36,
    },
    "type": "closeBrace",
    "value": "}",
  },
]
`;
